docs/source/config_mp.rst
=========
config_mp
=========

Data/geometry parameters
------------------------

- scan

| Mandatory, string type encapsulating scans or ranges of scans containing data for each peak. The scans/scan ranges should be arranged in ascending order.

::

    scan = "898-913,919-934,940-955,961-976"

- orientations

| Mandatory, a list of lists, each inner list defining the orientation of a peak.

::

    orientations = [[-1, -1, 1], [0, 0, 2], [1, 1, 1], [2, 0, 0]]

- hkl_in

| Mandatory, list with Miller indices representing the in-plane lattice vector.

::

    hkl_in = [3.031127677370978, 12.31353345906843, 8.75104158816168]

- hkl_out

| Mandatory, list with Miller indices representing the out-of-plane lattice vector.

::

    hkl_out = [9.77805918769193, -8.402719849515048, 8.436553021703112]

- twin_plane

| Mandatory, list with Miller indices of the twin plane. If there is not a twin plane, this should be the same as sample_axis.

::

    twin_plane = [1, -1, 1]

- sample_axis

| Mandatory, axis of the sample. The data will be rotated so that the twin plane vector in the lattice frame (hkl) corresponds to this vector in the array frame (xyz).

::

    sample_axis = [0, 1, 0]

- final_size

| Mandatory, a size in each dimension of the array holding reconstructed object.

::

    final_size = 180

- lattice_size

| Mandatory, lattice parameter of the reconstructing crystal. This is used to define the reciprocal lattice vectors, which are required for projecting to each peak.

::

    lattice_size = 0.4078

- rs_voxel_size
- ds_voxel_size

| Autofilled, These are calculated from the experimental geometry and added to the config file automatically. Do not edit these unless you REALLY know what you're doing.

General reconstruction parameters
---------------------------------

- switch_peak_trigger

| Mandatory, a trigger defining at which iteration to switch the peak.

::

    switch_peak_trigger = [5, 5]

- weight_init

| Mandatory, the initial global weight to use when updating the shared object. A weight of 0 will not update the shared object at all, while a weight of 1 will replace the shared object (except the orthogonal displacement) after phasing each peak. In general, a high weight leads to rapid development, but only forces the object to agree with whichever peak it phased most recently. A low weight converges slowly, but favors information common to all of the peaks.

::

    weight_init = 1.0

- weight_iters
- weight_vals

| Optional, list of iterations and corresponding list of values. After each iteration in weight_iters, the global weights will be updated to the corresponding value in weight_vals. A good rule of thumb is to start high and end low.

::

    weight_iters = [200, 400, 600, 800]
    weight_vals = [0.75, 0.5, 0.25, 0.1]

- calc_strain

| Optional, boolean, toggles whether to calculate the strain after reconstruction. Still experimental.

::

    calc_strain = False

Adaptive coupling parameters
----------------------------

- adapt_trigger

| Optional, determines when to update the weights assigned to individual peaks. Before each peak-switch, a confidence value is calculated for the current peak, based on how much it changed while phasing (i.e. how much this dataset disagrees with the current state of the reconstruction). These are then periodically compiled into a peak-specific weight, which determines how much each dataset is allowed to impact the reconstruction.

::

    adapt_trigger = [100, 100]

- adapt_power

| Optional, non-negative number that determines how harshly to punish bad datasets.

::

    adapt_power = 1  #

- adapt_threshold_init
- adapt_threshold_iters
- adapt_threshold_vals

| Optional, determines the relative confidence threshold required for shrinkwrap to prevent bad datasets from constantly derailing the reconstruction.

::

    adapt_threshold_init = 0.5
    adapt_threshold_iters = [200, 400, 600, 800]
    adapt_threshold_vals = [0.6, 0.7, 0.8, 0.9]

- adapt_alien_start

| Optional, determines when to begin adaptive alien removal. Alien removal occurs immediately after switching to a new peak, while the exit wave still agrees perfectly with the shared object. The diffraction amplitude calculated by forward propagating this object is compared with the actual measurement of the same reflection, and voxels where they strongly contradict each other are masked. A hybrid diffraction pattern, where the masked voxels have been replaced by the forward propagation, is then used for phasing.

::

    adapt_alien_start = 100

- adapt_alien_threshold

| Optional, determines the minimum amount of contradiction needed to mask a voxel. Under the hood, the actual value is this multiplied by the median of the normalized difference map.

::

    adapt_alien_threshold = 3


docs/source/config_rec.rst
.. _config_rec:

==========
config_rec
==========
| The "config_rec" file defines parameters used during reconstruction process.
|
| Parameters are grouped by trigger and corresponding parameters. There is a group of parameters that are not associated with any trigger, and thus, we call them 'general'.
| If trigger is defined then the feature is active and related parameters are used when processing the feature. Refer to :ref:`formula` page, for explanation of triggers, sub-triggers, and algorithm sequence.

Parameters
==========

General
+++++++
- data_dir

| Optional, defines a directory with file named "data.tif" that contains experiment data file ready for reconstruction. Default is <experiment_dir>/phasing_data.

::

    data_dir = "/path/to/phasing_data"

- save_dir

| Optional, used for specific cases, defualt is <experiment_dir>/results_phasing. Driven by the scripts, the experiment directory contains four directories: conf, preprocessed_data, phasing_data, results_phasing, and results_viz. The  reconstruction results are saved in <experiment_dir>/results_phasing directory. If the save_dir parameter is configured then the reconstruction result is saved in this directory.

::

    save_dir = "/path/to/results_phasing"

- init_guess

| Optional, defines how the initial guess is set. Possible options are: 'random', 'continue', and 'AI_guess'. The choice "random" will generate random guess, "continue" will start from previously saved results, and "AI_guess" will run AI reconstruction that will be an initial guess saved at <experiment_dir>/results_AI. Each of these algorithms require different parameters, explained below. The default is 'random'.

::

    init_guess = "random"

- continue_dir

| Must be defined if the init_guess parameter is set to 'continue'. Directory from which initial guess, initial support are read for reconstruction continuation. If the directory contains multiple subdirectories with initial guesses, a thread will start for each subdirectory.

::

    continue_dir = "/path/to/some_phasing_results_dir"

- AI_trained_model

| Must be defined, if init_guess is "AI_guess". Defines the file in hdf5 format that holds trained model.

::

    AI_trained_model = "/path/to/trained/model/trained_model.hdf5"

- reconstructions

| Optional, default is 1. Number of reconstructions to start with. Typically used when running genetic algorithm.

::

    reconstructions = 5

- processing

| Optional, the library used when running reconstruction. When the auto option is selected the program will use the best performing library that is available, in the following order: cupy, torch, numpy. The cp option will utilize cupy, torch will utilize torch, and np will utilize numpy. Default is auto.

::

    processing = "auto"

- device

| Optional, GPU IDs of the target devices for reconstruction(s) or 'all' if all available GPUs should be used. If not defined, the reconstruction process will run on CPU. For cluster configuration it is defined as dict with hosts names as keys and values as described.

::

    device = [0,1,2,7]
    device = 'all'
    device = {'host1':'all', 'host2':[0,1,2,3,4]}

- algorithm_sequence

| Mandatory, defines sequence of algorithms applied in each iteration during modulus projection and during modulus. The "*" character means repeat, and the "+" means add to the sequence. The sequence may contain single brackets defining a group that will be repeated by the preceding multiplier. The alphabetic entries: ER, ERpc, HIO, HIOpc define algorithms used in this iteration. The entries will invoke functions as follows: ER definition will invoke 'er' and 'modulus' functions, the ERpc will invoke 'er' and 'pc_modulus', HIO will invoke 'hio' and 'modulus', and HIOpc will invoke 'hio' and 'pc_modulus', 'RAAR' will invoke 'raar' and 'modulus', 'SF' will invoke 'sf' and 'modulus'. The pc_modulus is implementation of modulus with partial coherence correction. In second example the sequence contains sub-triggers, explained in  :ref:`formula` page.

::

    algorithm_sequence = "2* (20*SF + 180*HIO) + 2* (20*ERpc + 180*HIOpc) + 20*ERpc"
    algorithm_sequence = "20*ER.PHC0 + 180*HIO.PHC1 + 2* (20*ER.SW0 + 180*HIO.SW1) + 20*ER.SW2"

- hio_beta

| Optional, default is .9. A parameter used in hio algorithm.

::

    hio_beta = .9

- raar_beta

| Optional, default is .45. A parameter used in raar algorithm.

::

    raar_beta = .45

- initial_support_area

| Optional, defaults to [.5,.5,.5]. The list define dimensions of initial support area. The support area is calculated by multiplying the fractions in this parameter by the corresponding data array dimensions. The support array is centered.

::

    initial_support_area = [.5,.5,.5]

Twin
++++

- twin_trigger

| Defines at which iteration to eliminate "twin", i.e. the image under reconstruction is trimmed by preserving a quadrant across x and y dimensions and zeroing the rest of the array.

::

    twin_trigger = [2]

- twin_halves

| Optional, defines which quadrant of the array is preserved in x and y dimensions, defaults to (0, 0).
| Possible choices: [0, 0], [0, 1], [1, 0], [1,1]

::

    twin_halves = [0, 0]

Shrink wrap
+++++++++++
| Support area is an array that defines region in which the image is meaningful. This area is recalculated at the shrink wrap trigger iteration, shrinking along when the image develops. The calculations employ an algorithm defined here as shrink_wrap_type.

- shrink_wrap_trigger

| Defines when to update support array using the parameters below.
| Alternatively can be defined as list of sub-triggers. If sub-triggers are used, the parameters must be lists as well.

::

    shrink_wrap_trigger = [10, 1]
    shrink_wrap_trigger = [[10, 1],[0,5,100],[0,2]]   # sub-triggers

- shrink_wrap_type

| Mandatory, defines type of shrink wrap. Currently only the "GAUSS" type is supported that applies gaussian filter to find support area.

::

    shrink_wrap_type = "GAUSS"
    shrink_wrap_type = [GAUSS, GAUSS, GAUSS]  # sub-triggers

- shrink_wrap_threshold

| Mandatory, defines a threshold value used in the gaussian filter algorithm.

::

    shrink_wrap_threshold = 0.1
    shrink_wrap_threshold = [0.1, 0.11, .12]  # sub-triggers

- shrink_wrap_gauss_sigma

| Mandatory, defines a sigma value used in the gaussian filter algorithm.

::

    shrink_wrap_gauss_sigma = 1.0
    shrink_wrap_gauss_sigma = [1.0, 1.1, 1.2]  # sub-triggers

Phase constrain
+++++++++++++++
| At the beginning iterations the support area is modified in respect to the phase. Support area will include only points with calculated phase inside of the defined bounds.
| Alternatively can be defined as list of sub-triggers. If sub-triggers are used, the parameters must be lists as well.

- phc_trigger

| Defines when to update support array using the parameters below by applying phase constrain.

::

    phc_trigger = [0, 1, 310]
    phc_trigger = [[0, 1, 310], [0,2]]  # sub-triggers

- phc_phase_min

| Mandatory, defines lower bound phase.

::

    phc_phase_min = -1.57
    phc_phase_min = [-1.5, -1.57]  # sub-triggers

- phc_phase_max

| Mandatory, defines upper bound phase.

::

    phc_phase_max = 1.57
    phc_phase_max = [1.5, 1.57]  # sub-triggers

Partial coherence
+++++++++++++++++
| Partial coherence operation initiates recalculation of coherence of data with respect to the amplitudes in reciprocal space. After coherence array is determined, it is used in convolution operation during modulus in subsequent iteration. The coherence array is updated periodically, as defined by the pc_interval. Partial coherence operation is active if the interval is defined and the algorithm sequence indicates "pc".

- pc_interval

| Defines iteration interval between coherence update.

::

    pc_interval = 50

- pc_type

| Partial coherence algorithm. Currently "LUCY" is supported.

::

    pc_type = "LUCY"

- pc_LUCY_iterations

| Optional, defaults to 20. Defines number of iteration inside LUCY algorithm.

::

    pc_LUCY_iterations = 20

- pc_normalize

| Optional, defaults to True. Internal.

::

    pc_normalize = True

- pc_LUCY_kernel

| Mandatory, coherence array area.

::

    pc_LUCY_kernel = [16, 16, 16]

Lowpass Filter
++++++++++++++
| When active, a lowpass Gaussian filter is applied on data, with iteration dependent sigma calculated by line-spacing the lowpass_filter_range parameter over trigger span iterations. The low resolution trigger is typically configured to be active at the first part of iterations.

- lowpass_filter_trigger

| Defines when to apply lowpass filter operation using the parameters below. Typically the last trigger is configured at half of total iterations.

::

    lowpass_filter_trigger = [0, 1, 320]

- lowpass_filter_range

| The range is line-spaced over trigger iterations to form a list of iteration dependent sigmas. If only one number is given, the last sigma will default to 1.

::

    lowpass_filter_range = [.7, 1.0]

averaging
+++++++++
| When this feature is activated the amplitudes of the last several iterations are averaged.

- average_trigger

| Defines when to apply averaging. Negative start means it is offset from the last iteration.

::

    average_trigger = [-65, 1]

progress
++++++++
- progress_trigger

| Defines when to print info on the console. The info includes current iteration and error.

::

    progress_trigger = [0, 20]

live viewing
++++++++++++
| This feature allows for a live view of the amplitude, phase, support, and error as the reconstruction develops. With adaptive multipeak phasing, this will instead show the amplitude, phase, measured diffraction pattern, and adapted diffraction pattern. These are shown using a central slice cropped to half the full array size.

- live_trigger

| Defines when to update the live view.

::

    live_trigger = [0, 10]

GA
++
- ga_generations

| Defines number of generations. When defined, and the number is greater than 1, the genetic algorithm (GA) is activated

::

    ga_generations = 3

- ga_metrics

| Optional, a list of metrics that should be used to rank the reconstruction results for subsequent generations. If not defined, or shorter than number of generations, the metric defaults to "chi".
| If the list contains only one element, it will be used by all generations.
| Supported metrics:
| - 'chi': The last error calculated as norm(rs_amplitudes - data)/norm(data).
|           The smallest 'chi' value is the best.
| - 'sharpness': sum(power(abs(image), 4))
|           The smallest 'sharpness' value is the best.
| - 'summed_phase':  angle(image) - sum(angle(image) * support) / sum(support)
|           where support is calculated with shrink wrap using hardcoded threshold=.2 and sigma=.5
|           The greatest 'summed_phase' value is the best.
| - 'area': sum(support)
|           where support is calculated with shrink wrap using hardcoded threshold=.2 and sigma=.5
|           The greatest 'area' value is the best.

::

    ga_metrics = ["chi", "sharpness", "area"]
    ga_metrics = ["chi"]

- ga_breed_modes

| Optional, a list of breeding modes applied to breed consecutive generation. If not defined, or shorter that number of generations, the mode defaults to "sqrt_ab".
| If the list contains only one element, it will be used by all generations.
| Breeding starts with choosing alpha image. The rest of the images are crossed with alpha. Before the crossing, the image, called beta is aligned with alpha, and phases in both of the arrays are normalized to derive ph_alpha = angle(alpha), and ph_beta = angle(beta)
| Supported modes:
| - 'sqrt_ab': sqrt(abs(alpha) * abs(beta)) * exp(0.5j * (ph_beta + ph_alpha))
| - 'pixel_switch': where((cond > 0.5), beta, alpha); cond = random(shape(beta))
| - 'b_pa': abs(beta) * exp(1j * (ph_alpha))
| - '2ab_a_b': 2 * (beta * alpha) / (beta + alpha)
| - '2a_b_pa': (2 * abs(alpha) - abs(beta)) * exp(1j * ph_alpha)
| - 'sqrt_ab_pa': sqrt(abs(alpha) * abs(beta)) * exp(1j * ph_alpha)
| - 'sqrt_ab_recip': fftshift(ifft(fftshift(temp))), where temp is calculated below
|                      t1 = fftshift(fft(fftshift(beta)))
|                      t2 = fftshift(fft(fftshift(alpha)))
|                      temp = sqrt(abs(t1)*abs(t2))*exp(.5j*angle(t1))*exp(.5j*angle(t2))
| - 'max_ab': max(abs(alpha), abs(beta)) * exp(.5j * (ph_beta + ph_alpha))
| - 'max_ab_pa': max(abs(alpha), abs(beta)) * exp(1j * ph_alpha)
| - 'avg_ab': 0.5 * (alpha + beta)
| - 'avg_ab_pa: 0.5 * (abs(alpha) + abs(beta)) * exp(1j * (ph_alpha))

::

    ga_breed_modes = ["sqrt_ab", "pixel_switch", "none"]
    ga_breed_modes = ["sqrt_ab"]

- ga_cullings

| Optional, defines how many worst samples to remove in a breeding phase for each generation. If not defined for the generation, the culling defaults to 0.

::

    ga_cullings = [2,1]

- ga_sw_thresholds

| Optional, a list of threshold values for each generation. The support is recalculated with this threshold after breeding phase. Defaults to configured value of support_threshold.
| If the list contains only one element, it will be used by all generations.

::

    ga_sw_thresholds = [.15, .1]

- ga_sw_gauss_sigmas

| Optional, a list of sigma values for each generation. The support is recalculated with this sigma after breeding phase. Defaults to configured value of support_sigma.
| If the list contains only one element, it will be used by all generations.

::

    ga_sw_gauss_sigmas = [1.1, 1.0]

- ga_lpf_sigmas

| Optional, a list of sigmas that will be used in subsequent generations to calculate Gaussian low-pass filter applied it to the data. In the example given below this feature will be used in first two generations.

::

    ga_lpf_sigmas = [2.0, 1.5]

- ga_gen_pc_start

| Optional, a number indicating at which generation the partial coherence will start to be active. If not defined, and the pc feature is active, it will start at the first generation.

::

    ga_gen_pc_start = 3

- ga_fast

| Defines which GA algorithm to use. If present and True, the number of reconstructions is limited to available resources. This reconstruction is very fast. Otherwise the number of reconstructions is unlimited but the performance is worse as the intermediate results must be stored.

::

    ga_fast = True

cohere/controller/params.py
import pylibconfig2 as cfg
import os


class Params:
    def __init__(self, config_file):
        self.config_file = config_file

    def read_config(self):
        """
        This function gets configuration file. It checks if the file exists and parses it into an object.

        Parameters
        ----------
        config_file : str
            configuration file name, including path

        Returns
        -------
        config_map : Config object
            a Config containing parsed configuration, None if the given file does not exist
        """
        if os.path.isfile(self.config_file):
            try:
                with open(self.config_file, 'r') as f:
                    config_map = cfg.Config(f.read())
                    return (True, config_map)
            except Exception as e:
                msg = 'cannot parse configuration file', self.config_file, e
                return (False, msg)
        else:
            msg = 'configuration file', self.config_file, 'does not exist'
            return (False, msg)

    def set_params(self):
        (success, res) = self.read_config()
        if not success:
            return res
        conf = res

        if conf.lookup('algorithm_sequence') is None:
            return ('missing algorithm_sequence parameter in config file')
        if conf.lookup('data_dir') is not None:
            self.data_dir = conf.data_dir
        if conf.lookup('save_dir') is not None:
            self.save_dir = conf.save_dir

        if conf.lookup('init_guess') is not None:
            self.init_guess = conf.init_guess
            if self.init_guess == 'continue':
                if conf.lookup('continue_dir') is not None:
                    self.continue_dir = conf.continue_dir
                else:
                    return ('missing continue_dir parameter in config file')
            elif self.init_guess == 'AI_guess':
                if conf.lookup('AI_threshold') is not None:
                    self.AI_threshold = conf.AI_threshold
                else:
                    self.AI_threshold = conf.shrink_wrap_threshold
                if conf.lookup('AI_sigma') is not None:
                    self.AI_sigma = conf.AI_sigma
                else:
                    self.AI_sigma = conf.shrink_wrap_gauss_sigma
                if conf.lookup('AI_trained_model') is not None:
                    self.AI_trained_model = conf.AI_trained_model
            else:
                self.init_guess = 'random'
        else:
            self.init_guess = 'random'

        if conf.lookup('reconstructions') is not None:
            self.reconstructions = conf.reconstructions
        else:
            self.reconstructions = 1
        if conf.lookup('device') is not None:
            self.device = conf.device
        else:
            self.device = (-1)
        if conf.lookup('hio_beta') is not None:
            self.hio_beta = conf.hio_beta
        else:
            self.hio_beta = 0.9
        if conf.lookup('initial_support_area') is not None:
            self.initial_support_area = conf.initial_support_area
        else:
            self.initial_support_area = (.5, .5, .5)

        self.ga_gen_pc_start = None
        if conf.lookup('ga_generations') is not None:
            self.ga_generations = conf.ga_generations
        else:
            self.ga_generations = 1
        if self.ga_generations > 1:
            if conf.lookup('ga_fast') is not None:
                self.ga_fast = conf.ga_fast
            else:
                self.ga_fast = True
            if conf.lookup('self.metrics') is not None:
                self.metrics = list(conf.ga_metrics)
                if len(self.metrics) == 1:
                    self.metrics = self.metrics * self.ga_generations
                elif len(self.metrics) < self.ga_generations:
                    self.metrics = self.metrics + ['chi', ] * (self.ga_generations - len(self.metrics))
            else:
                self.metrics = ['chi', ] * self.ga_generations

            self.ga_reconstructions = []
            if conf.lookup('ga_cullings') is not None:
                worst_remove_no = list(conf.ga_cullings)
                if len(worst_remove_no) < self.ga_generations:
                    worst_remove_no = worst_remove_no + [0, ] * (self.ga_generations - len(worst_remove_no))
            else:
                worst_remove_no = [0, ] * self.ga_generations
            # calculate how many reconstructions should continue
            reconstructions = self.reconstructions
            for cull in worst_remove_no:
                reconstructions = reconstructions - cull
                if reconstructions <= 0:
                    return 'culled down to 0 reconstructions, check configuration'
                self.ga_reconstructions.append(reconstructions)

            if conf.lookup('ga_shrink_wrap_thresholds') is not None:
                self.ga_shrink_wrap_thresholds = list(conf.ga_shrink_wrap_thresholds)
                if len(self.ga_shrink_wrap_thresholds) == 1:
                    self.ga_shrink_wrap_thresholds = self.ga_shrink_wrap_thresholds * self.ga_generations
                elif len(self.ga_shrink_wrap_thresholds) < self.ga_generations:
                    if conf.lookup('shrink_wrap_threshold') is not None:
                        shrink_wrap_threshold = conf.shrink_wrap_threshold
                    else:
                        shrink_wrap_threshold = .1
                    self.ga_shrink_wrap_thresholds = self.ga_shrink_wrap_thresholds + [shrink_wrap_threshold, ] * (
                    self.ga_generations - len(self.ga_shrink_wrap_thresholds))
            else:
                if conf.lookup('shrink_wrap_threshold') is not None:
                    shrink_wrap_threshold = conf.shrink_wrap_threshold
                else:
                    shrink_wrap_threshold = .1
                self.ga_shrink_wrap_thresholds = [shrink_wrap_threshold, ] * (self.ga_generations)

            if conf.lookup('ga_shrink_wrap_gauss_sigmas') is not None:
                self.ga_shrink_wrap_gauss_sigmas = list(conf.ga_shrink_wrap_gauss_sigmas)
                if len(self.ga_shrink_wrap_gauss_sigmas) == 1:
                    self.ga_shrink_wrap_gauss_sigmas = self.ga_shrink_wrap_gauss_sigmas * self.ga_generations
                elif len(self.ga_shrink_wrap_gauss_sigmas) < self.ga_generations:
                    if conf.lookup('beta') is not None:
                        shrink_wrap_gauss_sigma = conf.shrink_wrap_gauss_sigma
                    else:
                        shrink_wrap_gauss_sigma = 1.0
                    self.ga_shrink_wrap_gauss_sigmas = self.ga_shrink_wrap_gauss_sigmas + [shrink_wrap_gauss_sigma, ] * (
                        self.ga_generations - len(self.ga_shrink_wrap_gauss_sigmas))
            else:
                if conf.lookup('shrink_wrap_gauss_sigma') is not None:
                    shrink_wrap_gauss_sigma = conf.shrink_wrap_gauss_sigma
                else:
                    shrink_wrap_gauss_sigma = 1.0
                self.ga_shrink_wrap_gauss_sigmas = [shrink_wrap_gauss_sigma, ] * (self.ga_generations)

            if conf.lookup('ga_breed_modes') is not None:
                self.breed_modes = list(conf.ga_breed_modes)
                if len(self.breed_modes) == 1:
                    self.breed_modes = self.breed_modes * self.ga_generations
                elif len(self.breed_modes) < self.ga_generations:
                    self.breed_modes = self.breed_modes + ['sqrt_ab', ] * (self.ga_generations - len(self.breed_modes))
            else:
                self.breed_modes = ['none', ] * self.ga_generations

            if conf.lookup('ga_lowpass_filter_sigmas') is not None:
                self.ga_lowpass_filter_sigmas = conf.ga_lowpass_filter_sigmas
                self.low_resolution_generations = len(self.ga_lowpass_filter_sigmas)
            else:
                self.low_resolution_generations = 0

            if self.low_resolution_generations > 0:
                if conf.lookup('low_resolution_alg') is not None:
                    self.low_resolution_alg = conf.ga_low_resolution_alg
                else:
                    self.low_resolution_alg = 'GAUSS'

            if conf.lookup('pc_interval') is not None:
                if conf.lookup('ga_gen_pc_start') is not None:
                    self.ga_gen_pc_start = conf.ga_gen_pc_start
                else:
                    self.ga_gen_pc_start = 0

        if conf.lookup('twin_trigger') is not None:
            if conf.lookup('twin_halves') is not None:
                self.twin_halves = conf.twin_halves
            else:
                self.twin_halves = (0, 0)

        self.shrink_wrap_gauss_sigma = None
        if conf.lookup('shrink_wrap_trigger') is not None:
            if conf.lookup('shrink_wrap_type') is not None:
                self.shrink_wrap_type = conf.shrink_wrap_type
            else:
                self.shrink_wrap_type = "GAUSS"
            if conf.lookup('shrink_wrap_threshold') is not None:
                self.shrink_wrap_threshold = conf.shrink_wrap_threshold
            else:
                self.shrink_wrap_threshold = 0.1
            if conf.lookup('shrink_wrap_gauss_sigma') is not None:
                self.shrink_wrap_gauss_sigma = conf.shrink_wrap_gauss_sigma
            else:
                self.shrink_wrap_gauss_sigma = 1.0

        if conf.lookup('phase_support_trigger') is not None:
            if conf.lookup('phm_phase_min') is not None:
                self.phm_phase_min = conf.phm_phase_min
            else:
                self.phm_phase_min = -1.57
            if conf.lookup('phm_phase_max') is not None:
                self.phm_phase_max = conf.phm_phase_max
            else:
                self.phm_phase_max = 1.57

        if conf.lookup('new_func_trigger') is not None:
            if conf.lookup('new_param') is not None:
                self.new_param = conf.new_param
            else:
                self.new_param = 1

        if conf.lookup('pc_interval') is not None:
            self.is_pcdi = True
            if conf.lookup('pc_type') is not None:
                self.pc_type = conf.pc_type
            else:
                self.pc_type = "LUCY"
            if conf.lookup('pc_LUCY_iterations') is not None:
                self.pc_LUCY_iterations = conf.pc_LUCY_iterations
            else:
                self.pc_LUCY_iterations = 20
            if conf.lookup('pc_normalize') is not None:
                self.pc_normalize = conf.pc_normalize
            else:
                self.pc_normalize = True
            if conf.lookup('pc_LUCY_kernel') is not None:
                self.pc_LUCY_kernel = conf.pc_LUCY_kernel
            else:
                self.pc_LUCY_kernel = (16, 16, 16)
        else:
            self.is_pcdi = False

        self.ll_sigmas = None
        self.ll_dets = None
        if conf.lookup('resolution_trigger') is not None and len(conf.resolution_trigger) == 3:
            # linespacing the sigmas and dets need a number of iterations
            # The trigger should have three elements, the last one
            # meaning the last iteration when the low resolution is
            # applied. If the trigger does not have the limit,
            # it is misconfigured, and the trigger will not be
            # active
            ll_iter = conf.resolution_trigger[2]
            if conf.lookup('shrink_wrap_trigger') is not None and conf.lookup('lowpass_filter_sw_sigma_range') is not None:
                # The sigmas are used to find support, if the shrink wrap
                # trigger is not active, it wil not be used
                sigma_range = conf.lowpass_filter_sw_sigma_range
                if len(sigma_range) > 0:
                    first_sigma = sigma_range[0]
                    if len(sigma_range) == 1:
                        last_sigma = self.shrink_wrap_gauss_sigma
                    else:
                        last_sigma = sigma_range[1]
                    self.ll_sigmas = [first_sigma + x * (last_sigma - first_sigma) / ll_iter for x in range(ll_iter)]
            self.ll_dets = None
            if conf.lookup('lowpass_filter_range') is not None:
                det_range = conf.lowpass_filter_range
                if len(det_range) > 0:
                    first_det = det_range[0]
                    if len(det_range) == 1:
                        last_det = 1.0
                    else:
                        last_det = det_range[1]
                    self.ll_dets = [first_det + x * (last_det - first_det) / ll_iter for x in range(ll_iter)]

        return None

